### ISSUE TITLE
Constantly Getting Stuck in a Local Minimum



### ISSUE THREAD

**Issue by emanuelbertey** [ZeroLikviz/godot_plugin-neural-network#6]
extends Control

# Datos de 121 carros -> [Antigüedad, costo de salida al mercado]

var x = [[0.0, 1.0], [0.1, 1.0], [0.2, 1.0], [0.3, 1.0], [0.4, 1.0],
	[0.5, 1.0], [0.6, 1.0], [0.7, 1.0], [0.8, 1.0], [0.9, 1.0],
	[1.0, 1.0], [0.0, 0.9], [0.1, 0.9], [0.2, 0.9], [0.3, 0.9],
	[0.4, 0.9], [0.5, 0.9], [0.6, 0.9], [0.7, 0.9], [0.8, 0.9],
	[0.9, 0.9], [1.0, 0.9], [0.0, 0.8], [0.1, 0.8], [0.2, 0.8],
	[0.3, 0.8], [0.4, 0.8], [0.5, 0.8], [0.6, 0.8], [0.7, 0.8],
	[0.8, 0.8], [0.9, 0.8], [1.0, 0.8], [0.0, 0.7], [0.1, 0.7],
	[0.2, 0.7], [0.3, 0.7], [0.4, 0.7], [0.5, 0.7], [0.6, 0.7],
	[0.7, 0.7], [0.8, 0.7], [0.9, 0.7], [1.0, 0.7], [0.0, 0.6],
	[0.1, 0.6], [0.2, 0.6], [0.3, 0.6], [0.4, 0.6], [0.5, 0.6],
	[0.6, 0.6], [0.7, 0.6], [0.8, 0.6], [0.9, 0.6], [1.0, 0.6],
	[0.0, 0.5], [0.1, 0.5], [0.2, 0.5], [0.3, 0.5], [0.4, 0.5],
	[0.5, 0.5], [0.6, 0.5], [0.7, 0.5], [0.8, 0.5], [0.9, 0.5],
	[1.0, 0.5], [0.0, 0.4], [0.1, 0.4], [0.2, 0.4], [0.3, 0.4],
	[0.4, 0.4], [0.5, 0.4], [0.6, 0.4], [0.7, 0.4], [0.8, 0.4],
	[0.9, 0.4], [1.0, 0.4], [0.0, 0.3], [0.1, 0.3], [0.2, 0.3],
	[0.3, 0.3], [0.4, 0.3], [0.5, 0.3], [0.6, 0.3], [0.7, 0.3],
	[0.8, 0.3], [0.9, 0.3], [1.0, 0.3], [0.0, 0.2], [0.1, 0.2],
	[0.2, 0.2], [0.3, 0.2], [0.4, 0.2], [0.5, 0.2], [0.6, 0.2],
	[0.7, 0.2], [0.8, 0.2], [0.9, 0.2], [1.0, 0.2], [0.0, 0.1],
	[0.1, 0.1], [0.2, 0.1], [0.3, 0.1], [0.4, 0.1], [0.5, 0.1],
	[0.6, 0.1], [0.7, 0.1], [0.8, 0.1], [0.9, 0.1], [1.0, 0.1],
	[0.0, 0.0], [0.1, 0.0], [0.2, 0.0], [0.3, 0.0], [0.4, 0.0],
	[0.5, 0.0], [0.6, 0.0], [0.7, 0.0], [0.8, 0.0], [0.9, 0.0],
	[1.0, 0.0]]

# 0 : normal    1 : coleccionable

var y = [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	1, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0,
	1, 1, 1, 0, 0, 0, 0, 0, 0, 0, 0,
	0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
	0, 1, 1, 1, 1, 1, 0, 0, 0, 0, 0,
	0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0,
	0, 0, 1, 1, 1, 1, 1, 0, 0, 0, 0,
	0, 0, 0, 1, 1, 1, 1, 1, 0, 0, 0,
	0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 0,
	0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0,
	0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1]
	
	

func _ready() -> void:
	print("Testing Simple XOR Networks with Multiple Algorithms:")
	test_xor_network()
	
	print("\nTest more complex networks with various algorithms and weighty initializations:")
	
	# Pruebas adicionales
	var configurations = [
		["Adam", "Xavier"],
		["NAG", "He"],
		["Nadam", "Random"],
		["Adamax", "Random"],
		["Adadelta", "Xavier"],
		["Yogi", "He"]
	]
	
	for config in configurations:
		print("Testing conf / test: " + config[0] + " with " + config[1] + " Initialization")
		test_complex_network([2, 10, 10, 1], config[0], config[1])
		test_complex_network([2, 50, 50, 1], config[0], config[1])
		test_complex_network([2, 10, 10, 10, 10, 1], config[0], config[1])

func test_xor_network() -> void:
	var nn = NNET.new([2, 5, 6, 1], true)
	nn.set_loss_function(BNNET.LossFunctions.MSE)
	nn.use_Rprop(0.3)
	nn.set_batch_size(4)

	run_training(nn, 1500, "XOR (Rprop)")
	XOR_test(nn)

func test_complex_network(architecture: Array, algorithm: String, init_method: String) -> void:
	var nn = NNET.new(architecture, false)
	nn.set_loss_function(BNNET.LossFunctions.MSE)
	set_algorithm(nn, algorithm)
	nn.set_batch_size(4)

	# Inicializar pesos
	match init_method:
		"Xavier":
			nn.reinit()
			print("Pesos inicializados con Xavier")
		"He":
			nn.reinit()
			print("Pesos inicializados con He")
		_:
			nn.reinit()  # Inicialización aleatoria por defecto

	run_training(nn, 1500, "Complex (" + algorithm + ", " + init_method + ")")
	XOR_test(nn)

func set_algorithm(nn: NNET, algorithm: String) -> void:
	match algorithm:
		"Adam":
			nn.use_Adam(0.001)
		"NAG":
			nn.use_NAG(0.1, 0.9)
		"Nadam":
			nn.use_Nadam(0.001)
		"Adamax":
			nn.use_Adamax(0.001)
		"Yogi":
			nn.use_Yogi(0.001)
		"Adadelta":
			nn.use_Adadelta(0.9)
		_:
			nn.use_Rprop(0.3)  # Algoritmo predeterminado

func run_training(nn: NNET, iterations: int, label: String) -> void:
	for i in range(iterations):
		
		var tag = []
		for k in range(y.size()):
			tag.append([y[k]])
		
		nn.train(x,tag)
		if i % 100 == 0:
			print(label, " entrenando la redes - Iteración ", i)
	print(label, "entrenamiento completo / Network Training Completed")

func XOR_test(nn: NNET) -> void:
	var select = randi() % 100
	var aux = []
	var aux2 = []
	aux.append(y[select])
	aux2.append([x[select]])
	print("----------------------------------------------------")
	print("Loss: ", nn.get_loss(x[select], aux))
	prints(x[select], " " , aux  , "  " , [y[select]] , "   " , x[select].size() )
	nn.set_input(x[select]); nn.propagate_forward()
	print(":" + str(y[select]), nn.get_output())
	nn.set_input([0.0, 0.9]); nn.propagate_forward()
	print("0.0, 0.9: ", nn.get_output())
	nn.set_input([0.1, 0.9]); nn.propagate_forward()
	print("0.1, 0.9: ", nn.get_output())
	nn.set_input([0.3, 1.0]); nn.propagate_forward()
	print("0.1, 0.9: ", nn.get_output())
	prints("se espera 0 , 1 , 1 , 0 : en al salida ")
	print("----------------------------------------------------")

I'm sorry, but it seems illogical to me that Loss asks you for a fixed-length output array at the input of the network, even if you encapsulate it in another array, check the set_input ??? it has to do with the exit with the entrance ??? I have to use fixed sizes ???  my array I protested it of 1 array and live input and the problem persists, also the tests here present demonstrate the linearity of the network, conclusion, it is not backward propagation or this very defective and it is only forward so it adjusts the first step weight and can do xor with a vector of 2,   If only two because of 3 or 4 already breaks, in this car example, you can't create a linear vector of x, and dimensions, are already calculated with Tesorflow, and it fails a lot by matching everything to 0.3 and something, which means it only averages the error as a whole and not per neuron per layer,  does the error of the last layer happen even to the first and you hide them averaging all the neurons equally ??' 

Attached proof

https://github.com/emagood/godot_plugin-neural-network/blob/main/prueba_dese.gd

I have a repo with proofs of this and more, it doesn't add up at all or I just don't know how to use this, although I really doubt it after many tests

update_momentums_adam() ,  before it worked fine without error, it broke in the last version

**Comment by ZeroLikviz** [comment]
Hi @emagood,

I looked at your code and I think now I understand what's happening.

1. The loss function (in my implementation) requires two arguments:

   - An array of input arrays
   - An array of desired output arrays

   If you want to calculate the loss for a single input array and a single target array, you need to wrap these arrays in additional outer arrays.


2. The coefficients you've chosen for the algorithms are far too small. Because of this, when the neural network falls into a local minimum, it doesn’t have enough momentum to escape and reach the global minimum (this is the "0.3, 0.3, 0.3" situation).


3. Your minibatch size is set to 4, which is too small for training on approximately 118 examples. I’ve changed it to 20, as using such a small batch size is like saying, "First go here, then go there, and don't bother remembering where I told you to go before - it doesn’t matter."



I’ve made some adjustments to your code, and it should work properly now. However, keep in mind that if your neural network still gets stuck in a local minimum, this is normal. Try modifying its structure, choosing a different algorithm, or, better yet, fine-tuning the algorithm’s coefficients.

```GDScript
extends Node

var x = [[0.0, 1.0], [0.1, 1.0], [0.2, 1.0], [0.3, 1.0], [0.4, 1.0],
		[0.5, 1.0], [0.6, 1.0], [0.7, 1.0], [0.8, 1.0], [0.9, 1.0],
		[1.0, 1.0], [0.0, 0.9], [0.1, 0.9], [0.2, 0.9], [0.3, 0.9],
		[0.4, 0.9], [0.5, 0.9], [0.6, 0.9], [0.7, 0.9], [0.8, 0.9],
		[0.9, 0.9], [1.0, 0.9], [0.0, 0.8], [0.1, 0.8], [0.2, 0.8],
		[0.3, 0.8], [0.4, 0.8], [0.5, 0.8], [0.6, 0.8], [0.7, 0.8],
		[0.8, 0.8], [0.9, 0.8], [1.0, 0.8], [0.0, 0.7], [0.1, 0.7],
		[0.2, 0.7], [0.3, 0.7], [0.4, 0.7], [0.5, 0.7], [0.6, 0.7],
		[0.7, 0.7], [0.8, 0.7], [0.9, 0.7], [1.0, 0.7], [0.0, 0.6],
		[0.1, 0.6], [0.2, 0.6], [0.3, 0.6], [0.4, 0.6], [0.5, 0.6],
		[0.6, 0.6], [0.7, 0.6], [0.8, 0.6], [0.9, 0.6], [1.0, 0.6],
		[0.0, 0.5], [0.1, 0.5], [0.2, 0.5], [0.3, 0.5], [0.4, 0.5],
		[0.5, 0.5], [0.6, 0.5], [0.7, 0.5], [0.8, 0.5], [0.9, 0.5],
		[1.0, 0.5], [0.0, 0.4], [0.1, 0.4], [0.2, 0.4], [0.3, 0.4],
		[0.4, 0.4], [0.5, 0.4], [0.6, 0.4], [0.7, 0.4], [0.8, 0.4],
		[0.9, 0.4], [1.0, 0.4], [0.0, 0.3], [0.1, 0.3], [0.2, 0.3],
		[0.3, 0.3], [0.4, 0.3], [0.5, 0.3], [0.6, 0.3], [0.7, 0.3],
		[0.8, 0.3], [0.9, 0.3], [1.0, 0.3], [0.0, 0.2], [0.1, 0.2],
		[0.2, 0.2], [0.3, 0.2], [0.4, 0.2], [0.5, 0.2], [0.6, 0.2],
		[0.7, 0.2], [0.8, 0.2], [0.9, 0.2], [1.0, 0.2], [0.0, 0.1],
		[0.1, 0.1], [0.2, 0.1], [0.3, 0.1], [0.4, 0.1], [0.5, 0.1],
		[0.6, 0.1], [0.7, 0.1], [0.8, 0.1], [0.9, 0.1], [1.0, 0.1],
		[0.0, 0.0], [0.1, 0.0], [0.2, 0.0], [0.3, 0.0], [0.4, 0.0],
		[0.5, 0.0], [0.6, 0.0], [0.7, 0.0]]

var y = [[0], [0], [0], [0], [0], [0], [0], [0], [0], [0], [0],
		 [1], [1], [0], [0], [0], [0], [0], [0], [0], [0], [0],
		 [1], [1], [1], [0], [0], [0], [0], [0], [0], [0], [0],
		 [0], [1], [1], [1], [1], [1], [0], [0], [0], [0], [0],
		 [0], [1], [1], [1], [1], [1], [0], [0], [0], [0], [0],
		 [0], [0], [1], [1], [1], [1], [1], [0], [0], [0], [0],
		 [0], [0], [1], [1], [1], [1], [1], [0], [0], [0], [0],
		 [0], [0], [0], [1], [1], [1], [1], [1], [0], [0], [0],
		 [0], [0], [0], [1], [1], [1], [1], [1], [1], [1], [0],
		 [0], [0], [0], [0], [1], [1], [1], [1], [1], [1], [0],
		 [0], [0], [0], [0], [0], [0], [1], [1]]



func _ready() -> void:
		var configurations = [
				["Adam"],
				["NAG"],
				["Nadam"],
				["Adamax"],
				["Adadelta"],
				["Yogi"],
				["Rprop"]
		]

		for config in configurations:
				print("Testing configuration " + config[0])
				test_complex_network([2, 10, 10, 1], config[0])
				test_complex_network([2, 10, 10, 10, 10, 1], config[0])

func test_complex_network(architecture: Array, algorithm: String) -> void:
		var nn = NNET.new(architecture, false)
		nn.set_loss_function(BNNET.LossFunctions.MSE)
		set_algorithm(nn, algorithm)
		nn.set_batch_size(20)

		run_training(nn, 3500, "Complex (" + algorithm + ")")
		show_results(nn)

func set_algorithm(nn: NNET, algorithm: String) -> void:
		match algorithm:
				"Adam":
						nn.use_Adam(0.1)
				"NAG":
						nn.use_NAG(0.1, 0.9)
				"Nadam":
						nn.use_Nadam(0.1)
				"Adamax":
						nn.use_Adamax(0.1)
				"Yogi":
						nn.use_Yogi(0.1)
				"Adadelta":
						nn.use_Adadelta(0.5)
				"Rprop":
						nn.use_Rprop(0.3)

func run_training(nn: NNET, iterations: int, label: String) -> void:
		for i in range(iterations):
				nn.train(x,y)
				if i % 500 == 0:
						print(label, "Training the network. Iteration number ", i)
		print(label, "Network training is completed")

func show_results(nn: NNET) -> void:
		print(nn.algorithm_to_string(nn.algorithm), ": ", nn.structure)
		print("Loss: ", nn.get_loss(x, y))
		nn.set_input([0.0, 0.9]); nn.propagate_forward()
		print("0.0, 0.9: ", nn.get_output())
		nn.set_input([0.1, 0.9]); nn.propagate_forward()
		print("0.1, 0.9: ", nn.get_output())
		nn.set_input([0.3, 1.0]); nn.propagate_forward()
		print("0.3, 1.0: ", nn.get_output())
		print("----------------------------------------------------")

```

Here are the results I got:
```
Adam: [2, 10, 10, 1]
Loss: 0.03689584748238
0.0, 0.9: [0.94948755066073]
0.1, 0.9: [0.81415316790675]
0.3, 1.0: [0.01089840006109]
----------------------------------------------------
Adam: [2, 10, 10, 10, 10, 1]
Loss: 0.01743326088354
0.0, 0.9: [0.98891059589344]
0.1, 0.9: [0.90319446873362]
0.3, 1.0: [0.00043424391682]
----------------------------------------------------
NAG: [2, 10, 10, 1]
Loss: 0.22322665065544
0.0, 0.9: [0.37062614266631]
0.1, 0.9: [0.36100057659254]
0.3, 1.0: [0.32062250208624]
----------------------------------------------------
NAG: [2, 10, 10, 10, 10, 1]
Loss: 0.23217277386942
0.0, 0.9: [0.40404679284043]
0.1, 0.9: [0.39941014013516]
0.3, 1.0: [0.38688146359599]
----------------------------------------------------
Nadam: [2, 10, 10, 1]
Loss: 0.03491609482231
0.0, 0.9: [0.76634914920899]
0.1, 0.9: [0.67879040766608]
0.3, 1.0: [0.01669019475836]
----------------------------------------------------
Nadam: [2, 10, 10, 10, 10, 1]
Loss: 0.03005831430888
0.0, 0.9: [0.97458830283031]
0.1, 0.9: [0.48318143929745]
0.3, 1.0: [0.02176581835142]
----------------------------------------------------
Adamax: [2, 10, 10, 1]
Loss: 0.08892553863336
0.0, 0.9: [0.43950933925129]
0.1, 0.9: [0.5609222258008]
0.3, 1.0: [0.00825470346455]
----------------------------------------------------
Adamax: [2, 10, 10, 10, 10, 1]
Loss: 0.04629364003237
0.0, 0.9: [0.94302904592397]
0.1, 0.9: [0.13986608243846]
0.3, 1.0: [0.02324446388384]
----------------------------------------------------
Adadelta: [2, 10, 10, 1]
Loss: 0.08203367800326
0.0, 0.9: [0.6556955647339]
0.1, 0.9: [0.63839855662702]
0.3, 1.0: [0.0225445274215]
----------------------------------------------------
Adadelta: [2, 10, 10, 10, 10, 1]
Loss: 0.06365105749481
0.0, 0.9: [0.95396479776194]
0.1, 0.9: [0.88100212406305]
0.3, 1.0: [0.03870116568529]
----------------------------------------------------
Yogi: [2, 10, 10, 1]
Loss: 0.07139098732351
0.0, 0.9: [0.71170647823672]
0.1, 0.9: [0.77250789969571]
0.3, 1.0: [0.15888958991044]
----------------------------------------------------
Yogi: [2, 10, 10, 10, 10, 1]
Loss: 0.15021647756649
0.0, 0.9: [0.60210051465488]
0.1, 0.9: [0.59826104850168]
0.3, 1.0: [0.00030428459706]
----------------------------------------------------
Rprop: [2, 10, 10, 1]
Loss: 0.23604224791419
0.0, 0.9: [0.34852698886467]
0.1, 0.9: [0.34539266155032]
0.3, 1.0: [0.33858516308813]
----------------------------------------------------
Rprop: [2, 10, 10, 10, 10, 1]
Loss: 0.19968791040515
0.0, 0.9: [0.35137476578278]
0.1, 0.9: [0.31191299403975]
0.3, 1.0: [0.2027110199454]
```

I hope this answers all your questions. If you still have any, feel free to ask me.
